get-terraform-provider-versions/get_terraform_provider_versions/get-tf-provider-versions.py

```diff
@@ -0,0 +1,337 @@
+#!/usr/bin/env python3
+import sys
+import getopt
+import os
+import re
+import logging
+import tempfile
+import subprocess
+import requests
+
+from github.repo import Repo
+
+GITHUB_ORGANIZATION = 'dfds'
+TFREGISTRY_BASEAPI = 'https://registry.terraform.io/v1/providers/'
+
+
+# define a custom class to hold the providers we locate
+class TerraformProvider:
+    """
+    This class defines a Terraform provider
+    """
+
+    def __init__(self, repository_name: str, file_path: str, provider_name: str, provider_version: str, latest_provider_version: str = "", comment: str = ""):
+        self.repository_name = repository_name
+        self.file_path = file_path
+        self.provider_name = provider_name
+        self.provider_version = provider_version
+        self.latest_provider_version = latest_provider_version
+        self.comment = comment
+
+
+class TerraformProviders:
+    """
+    This class defines a collection of Terraform providers
+    """
+
+    def __init__(self):
+        self.terraform_providers: list = []
+
+    def __iter__(self):
+        return iter(self.terraform_providers)
+
+    def append(self, new_item):
```

Comment:
This method if use will have a huge performance impact. Looping over collections, and then adding it one by one to a class variable is so slow it is discouraged in https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Local_Variables 

It is better to do this looping and appending in a local context, and when done, then call a class method like 

```
set_terraform_providers(self, new_list) -> None:
  self.terraform_providers = new_list
```

Reply:
I don't use __iter__ and append together.

So far as I can see reading back through my code the iterator is only included to provide me with a means to loop through the collection when building the output string (either CSV, JSON or Table).

The append is called when a new instance of the TerraformProvider class is created and this is driven by some parsing logic in the loop that reads the file.

I totally understand your point but I don't think it's relevent in this case.   If i'm wrong then please let me know :-)